### DISCUSSION TITLE
Are we recycling the bitmaps in the transformation chain?

### DISCUSSION THREAD

**Discussion by halizton** [coil-kt/coil#1309] [Q&A]
While reading through the code, I did not seem to find that we are recycling the intermediate bitmaps created in the transformation chain. We are migrating from using Picasso to using Coil and experienced increased OOM issues. We are using resize transformation and I am wondering if the lack of bitmap recycling would contribute to the increased OOM issues.

coil version: master rev ea2ccd8fb0458ff090b6f7adf88d427775f87250
While searching the bitmap.recycle() method, I found 3 usages throughout the code base

![image](https://user-images.githubusercontent.com/12901839/171908702-6b59d1db-7b10-4fcf-b54b-f056ca7222b3.png)

**Comment by colinrtwhite** [comment]
No, Coil only recycles bitmaps that aren't exposed in the public API. Iirc Picasso doesn't recycles any bitmaps. Instead of using a resize transformation you should specify the size using `ImageRequest.Builder.size`.

**Comment by halizton** [comment]
As I mentioned earlier in https://github.com/coil-kt/coil/issues/1128, but the exact size won't be respected.

I think Picasso does recycle the intermediate bitmap used for transformation
https://github.com/Pixate/picasso/blob/master/picasso/src/main/java/com/squareup/picasso/BitmapHunter.java#L471

> No, Coil only recycles bitmaps that aren't exposed in the public API.

Ah, I see, do you mean the the `input` bitmap is exposed in the public API, so it is the Transformation implementation's responsibility to recycle the `input` bitmap and return the new `bitmap` to the pipeline?
`Transformation::transform(input: Bitmap, size: Size): Bitmap`
